### ISSUE TITLE
Make same number of random stream calls regardless of code path

### ISSUE THREAD

**Issue by ncollier** [NSF-RESUME/covid-age#5]
Update the code so that we always make the same number of calls to the random stream regardless of the code path. 

* Create a struct that implements the UniformRandomBitGenerator interface.
* Fill that object with N (100) number of "bit generations" from the existing random number generator
* Use that object in place of the existing rng.
* If the objects cached values become exhausted, start over with the first, and print a warning  - "Use a bigger N".

The new generator should be created and filled immediately prior to this line: https://github.com/NSF-RESUME/covid-age/blob/1609490b06cfdbb1ad832209ff0ad48d84e789dd/src/NUCOVID_cereal.h#L534.

Use it in place of the rng there, and the `infect` call needs to be updated to take the generator object as the rng for the rand_X calls made in `infect`. 

See https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator for more on the uniform random bit generation interface.

**Comment by ncollier** [comment]
The above doesn't seem to solve the issue. So some further debugging.

Setting the CachedBitGenerator to always return only the same single value, and running

```bash
./model '{"random_seeds":[[0,1111]],"ini_Ki":1,"Ki_ap":[[0,0.4],[34,0.3]], "duration":32, "output_filename": "03_output.txt"}'
./model '{"random_seeds":[[0,1111]],"ini_Ki":1,"Ki_ap":[[0,0.4],[34,0.5]], "duration":32, "output_filename": "05_output.txt"}'
```

yields a diff with:

```bash
nick@ubuntu-22041:~/Documents/repos/covid-age/exp/chicago_yr1$ diff 03_output.txt 05_output.txt 
23,25c23,25
< 0     30      0.4     2499977 12      10      1       0       0       0       0       1       0       0
< 0     31      0.4     2499971 14      12      2       0       0       0       1       3       0       0
< 0     32      0.4     2499952 29      11      5       0       0       0       3       8       0       0
---
> 0     30      0.4     2499979 10      11      0       0       0       0       0       0       0       0
> 0     31      0.4     2499973 14      11      1       0       0       0       1       2       0       0
> 0     32      0.4     2499969 17      11      2       0       0       0       1       3       0       0
```

**Comment by ncollier** [comment]
Issue may be in here: https://github.com/NSF-RESUME/covid-age/blob/55268092bad05624e007c9600052c0411fd0f223/src/NUCOVID_cereal.h#L491

```c++
double Tc = rand_exp(n->get_Ki((int) Ti) * Ki_modifier[bin], &cbg) + Ti;
while ( Tc < Tr ) {     // does contact occur before recovery?
    // decide which node to infect
    size_t infect_node_id = get_infection_node_id(n->id);
    add_event(Tc, CON, n, nodes[infect_node_id], det_flag); // potential transmission event
    while (bin < Times.size() - 1 and Times[bin+1] < Tc) {bin++;} // update bin if necessary
    Tc += rand_exp(n->get_Ki((int) Tc) * Ki_modifier[bin], &cbg);
}
```

If Ti is >= 34 then n->get_Ki will return a different value in the two scenarios. Consequently, Tc will be different and the draws from cbg are then out of sync. I confirmed that Ti is >= 34 in some instances and the number of times the contents of this loop are executed does diverge between the two.

**Comment by ncollier** [comment]
Fixed in 84e037e
